analytics_data_api/utils.py

```diff
@@ -84,3 +112,113 @@ def date_range(start_date, end_date, delta=datetime.timedelta(days=1)):
+        # - Such a method isn't implemented
+        # - We can't import the specified storage class
+        # - We don't have privileges for the specified file location
+        raise CannotCreateReportDownloadLinkError
+
+    try:
+        last_modified = default_storage.modified_time(report_location)
+    except (NotImplementedError, AttributeError):
+        last_modified = None
+
+    try:
+        download_size = default_storage.size(report_location)
+    except (NotImplementedError, AttributeError):
+        download_size = None
+
+    download_filename = '{}-{}-{}.csv'.format(
+        course_id,
+        report_name,
+        # We need a date for the filename; if we don't know when it was last modified,
+        # use the current date and time to stamp the filename.
+        (last_modified or datetime.datetime.utcnow()).strftime('%Y%m%dT%H%M%SZ')
+    )
+    url, expiration_date = get_file_object_url(report_location, download_filename)
+
+    details = {
+        'course_id': course_id,
+        'report_name': report_name,
+        'download_url': url
+    }
+    # These are all optional items that aren't guaranteed. The URL isn't guaranteed
+    # either, but we'll raise an exception earlier if we don't have it. Currently, support
+    # is only present for S3, which has all these data types, but this future-proofs
+    # us for scenarios where we might have a storage provider that doesn't support them.
+    if last_modified is not None:
+        details.update({'last_modified': last_modified.strftime(settings.DATETIME_FORMAT)})
+    if expiration_date is not None:
+        details.update({'expiration_date': expiration_date.strftime(settings.DATETIME_FORMAT)})
+    if download_size is not None:
+        details.update({'file_size': download_size})
+    return details
+
+
+def get_file_object_url(filename, download_filename):
+    """
+    Retrieve a download URL for the file, as well as a datetime object
+    indicating when the URL expires.
+
+    We need to pass extra details to the URL method, above and beyond just the
+    file location, to give us what we need.
+
+    Currently, this method only supports S3 storage, and will need to be changed to
+    support building URLs on other storage providers.
+    """
+    # Default to expiring the link after two minutes
+    expire_length = getattr(settings, 'COURSE_REPORT_DOWNLOAD_EXPIRY_TIME', 120)
+    expires_at = get_expiration_date(expire_length)
+    try:
+        url = default_storage.url(
+            name=filename,
+            response_headers={
+                'response-content-disposition': 'attachment; filename={}'.format(download_filename),
+                'response-content-type': 'text/csv',
+                # The Expires header requires a very particular timestamp format
+                'response-expires': expires_at.strftime('%a, %d %b %Y %H:%M:%S GMT')
+            },
+            expire=expire_length
+        )
```

Comment:
@haikuginger So, I know we said at the beginning of this task that we were only going to support S3 files, and you did that, which was great.

But @bradenmacdonald raised the point during his testing that he's never had to set up S3 storage for the analytics devstack before, and now @maxrothman is referring to [supporting the default filesystem storage](https://github.com/edx/configuration/pull/3286#issuecomment-239831593) in the configuration.

So I think we should support the default filesystem storage in dev, and allow S3 to be used in production.  

To help things along, I've tested the following code and configuration on my devstack, and it works, though of course doesn't have the nice expiry and timestamped download filename that S3 allows.

`analytics_data_api/url.py` diff:

```diff
-    Currently, this method only supports S3 storage, and will need to be changed to
-    support building URLs on other storage providers.
+    Currently, this method only supports FileSystemStorage and S3 storage, and 
+    will need to be changed to support building URLs on other storage providers.
     """
-    # Default to expiring the link after two minutes
-    expire_length = getattr(settings, 'COURSE_REPORT_DOWNLOAD_EXPIRY_TIME', 120)
-    expires_at = get_expiration_date(expire_length)
-    try:
-        url = default_storage.url(
-            name=filename,
+    url_params = dict(name=filename)
+    expires_at = None
+
+    expire_length = getattr(settings, 'COURSE_REPORT_DOWNLOAD_EXPIRY_TIME', None)
+    if expire_length is not None:
+        expires_at = get_expiration_date(expire_length)
+        url_params += dict(
             response_headers={
                 'response-content-disposition': 'attachment; filename={}'.format(download_filename),
                 'response-content-type': 'text/csv',
@@ -210,6 +216,8 @@ def get_file_object_url(filename, download_filename):
             },
             expire=expire_length
         )
+    try:
+        url = default_storage.url(**url_params)
     except (AttributeError, TypeError, NotImplementedError):
         # Either we can't find a .url() method, or we can't use it. Raise an exception.
         raise CannotCreateReportDownloadLinkError
```


configuration:  append to `analyticsdataserver/settings/base.py`

```yaml
REPORT_DOWNLOAD_BACKEND:
  DEFAULT_FILE_STORAGE: 'django.core.files.storage.FileSystemStorage'
  MEDIA_ROOT: '/edx/app/analytics_api/analytics_api/static/reports'
  MEDIA_URL: 'http://localhost:8100/static/reports/'   # note absolute URL, and trailing slash
  COURSE_REPORT_FILE_LOCATION_TEMPLATE: '{course_id}_{report_name}.csv'   # note no leading slash
```